src/glum/_transformers.py

```diff
@@ -0,0 +1,347 @@
+    drop_idx = np.where(~keep_mask)[0]
+
+    intercept_safe = False
+    if fit_intercept:
+        if 0 not in drop_idx:
+            intercept_safe = True
+        keep_idx -= 1
+        drop_idx -= 1
+
+    keep_idx = keep_idx[keep_idx != -1]
+    drop_idx = drop_idx[drop_idx != -1]
+
+    return CollinearityResults(keep_idx, drop_idx, intercept_safe)
+
+
+class ColumnMap(NamedTuple):
+    """Mapping from DataFrame to design matrix."""
+
+    column_pos: int
+    column_name: Hashable
+    categorical: bool
+    category: Optional[str] = None
+    base_category: Optional[str] = None
+
+
+def _get_column_mapping(X: pd.DataFrame) -> Sequence[ColumnMap]:
+    column_mapping = []
+    for column_pos, (column_name, dtype) in enumerate(X.dtypes.items()):
+        if isinstance(dtype, pd.CategoricalDtype):
+            if len(dtype.categories) > 1:
+                base_category = dtype.categories[0]
+                for category in dtype.categories[1:]:
+                    column_mapping.append(
+                        ColumnMap(
+                            column_pos, column_name, True, category, base_category
+                        )
+                    )
+        else:
+            column_mapping.append(ColumnMap(column_pos, column_name, False))
+    return column_mapping
+
+
+class Decollinearizer(TransformerMixin, BaseEstimator):
+    """Drop collinear columns from the design matrix implied by a dataset.
+
+    The type of the output is the same as the input. For non-categorical
+    columns, collinear ones are simply dropped. For categorical columns
+    (e.g. in a pandas.DataFrame or a tabmat.SplitMatrix), values whose
+    columns in the design matrix would be dropped are replaced with the
+    first category. This supposes that the first category is the reference,
+    and will be dropped in the subsequent model fitting step.
+    """
+
+    drop_columns: Union[Sequence[Hashable], Sequence[int]]
+    keep_columns: Union[Sequence[Hashable], Sequence[int]]
+    intercept_safe: bool
+    input_type: str
+    replace_categories: List[Tuple[Hashable, str, str]]
+
+    def __init__(self, fit_intercept: bool = True, tolerance: float = 1e-6) -> None:
+        self.fit_intercept = fit_intercept
+        self.tolerance = tolerance
+
+    def fit(
+        self,
+        X: ArrayLike,
+        y: Optional[VectorLike] = None,
+        mode: Optional[str] = None,
+        use_tabmat: Optional[bool] = None,
+    ) -> "Decollinearizer":
+        """Fit the transformer by finding a maximal set of linearly independent columns.
+
+        Parameters
+        ----------
+        X : ArrayLike
+            The data to fit. Can be a pandas.DataFrame
+            a tabmat.SplitMatrix, or a numpy.ndarray.
+        y: Optional[VectorLike]
+            Ignored. Present for API consistency.
+
+        Returns
+        -------
+        Self
+            The fitted transformer.
+        """
+        if mode not in ["gram", "direct", None]:
```

Comment:
It would be useful to add a test that the product of modes QR and gram and numpy and tabmat ways all give the same result. I would also be happy with reducing the number of options in case there is a favorite. E.g, if gram works well everywhere and is faster for `k << n`, then we can scrap the direct mode.

Reply:
Agreed. I'll create some tests that check the modes against each other. I'll also do some benchmarks and then we can decide how to proceed.